clippy_lints/src/returned_iter_not_implementing_iterator.rs

```diff
@@ -0,0 +1,112 @@
+use clippy_utils::get_trait_def_id;
+use clippy_utils::return_ty;
+use clippy_utils::{paths, ty::implements_trait};
+use rustc_hir::ImplItem;
+use rustc_hir::ImplItemKind;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Adt;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// detecting methods `iter_mut()` and `iter()` returning a type not implementing 'Iterator'
+    /// ### Why is this bad?
+    /// methods `iter_mut()` and `iter()` are expected to return Iterators, and Iterators are expected to implement trait `Iterator`
+    /// ### Example
+    /// ```rust
+    /// //Counter didn't implement `Iterator`
+    /// struct Data {
+    ///     begin: u32,
+    /// }
+    ///
+    /// struct Counter {
+    ///     count: u32,
+    /// }
+    ///
+    /// impl Counter {
+    ///     fn new(x: u32) -> Counter {
+    ///         Counter { count: x }
+    ///     }
+    /// }
+    ///
+    /// impl Data {
+    ///     fn iter(self) -> Counter {
+    ///         Counter::new(self.begin)
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct Data {
+    ///     begin: u32,
+    /// }
+    ///
+    /// struct Counter {
+    ///     count: u32,
+    /// }
+    ///
+    /// impl Counter {
+    ///     fn new(x: u32) -> Counter {
+    ///         Counter { count: x }
+    ///     }
+    /// }
+    ///
+    /// impl Data {
+    ///     fn iter(self) -> Counter {
+    ///         Counter::new(self.begin)
+    ///     }
+    /// }
+    ///
+    /// impl Iterator for Counter {
+    ///     type Item = u32;
+    ///
+    ///     fn next(&mut self) -> Option<Self::Item> {
+    ///         self.count += 1;
+    ///
+    ///         if self.count < 6 {
+    ///             Some(self.count)
+    ///         } else {
+    ///             None
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    pub RETURNED_ITER_NOT_IMPLEMENTING_ITERATOR,
+    restriction,
+    "detecting `iter()` and `iter_mut()` returning type not implementing `Iterator`"
+}
+
+declare_lint_pass!(ReturnedIterNotImplementingIterator => [RETURNED_ITER_NOT_IMPLEMENTING_ITERATOR]);
+
+impl LateLintPass<'_> for ReturnedIterNotImplementingIterator {
+    fn check_impl_item(&mut self, cx: &LateContext<'tcx>, impl_item: &'tcx ImplItem<'tcx>) {
+        let name: &str = &impl_item.ident.name.as_str();
+        let method_names = vec!["iter", "iter_mut"];
+        if_chain! {
+            if let ImplItemKind::Fn(fn_sig, _) = &impl_item.kind;
+            let ret_ty = return_ty(cx, impl_item.hir_id());
+            if let Adt(adt_def, _) = ret_ty.kind();
+            if adt_def.is_struct();
+            if method_names.contains(&name);
+
+            then {
+                let parameters = fn_sig.decl.inputs;
+                if parameters.is_empty() || parameters.len() > 1 {
+                    return;
+                }
```

Comment:
Why check for number of parameters?
Shouldn't this lint work for an iter that takes parameters other than self?
Or maybe this is covered by another lint?

Reply:
I think it should only detect `fn iter(&self) -> _` before, because I think if it has more parameters it may have a different meaning, so this lint may not suit it.